src/main/java/dev/openfeature/sdk/OpenFeatureClient.java

```diff
@@ -183,17 +182,29 @@ private static <T> void enrichDetailsWithErrorDefaults(T defaultValue, FlagEvalu
      * @return merged evaluation context
      */
     private EvaluationContext mergeEvaluationContext(EvaluationContext invocationContext) {
-        final EvaluationContext apiContext = openfeatureApi.getEvaluationContext() != null
-                ? openfeatureApi.getEvaluationContext()
-                : new ImmutableContext();
-        final EvaluationContext clientContext = this.getEvaluationContext() != null
-                ? this.getEvaluationContext()
-                : new ImmutableContext();
-        final EvaluationContext transactionContext = openfeatureApi.getTransactionContext() != null
-                ? openfeatureApi.getTransactionContext()
-                : new ImmutableContext();
-
-        return apiContext.merge(transactionContext.merge(clientContext.merge(invocationContext)));
+        // avoid any unnecessary context instantiations and stream usage here; this is call with every evaluation.
```

Comment:
This method is way uglier than before with a nice nested call to `.merge()` but this results in way less memory usage.

Reply:
As I see it, we're creating a new `ImmutableContext` for each `merge()` invocation. It seems like we could optimize this a bit more by skipping the merging of empty contexts. How do you feel about that? Here's an example of what I have in mind:
```
        final EvaluationContext apiContext = openfeatureApi.getEvaluationContext();
        final EvaluationContext clientContext = this.getEvaluationContext();
        final EvaluationContext transactionContext = openfeatureApi.getTransactionContext();
        final List<EvaluationContext> contextsToMerge = new ArrayList<>();
        contextsToMerge.add(apiContext);
        contextsToMerge.add(transactionContext);
        contextsToMerge.add(clientContext);
        contextsToMerge.add(invocationContext);

        EvaluationContext merged = new ImmutableContext();
        for (EvaluationContext evaluationContext : contextsToMerge) {
            if (evaluationContext != null && !evaluationContext.isEmpty()) {
                merged = merged.merge(evaluationContext);
            }
        }
        return merged;
```

Reply:
But even if we do so, in the worst case (all contexts are not empty), we create 5 `ImmutableContext` objects per invocation. 
What about pulling the merge logic out into a factory method on the 'ImmutableContext' that accepts 4 named contexts to merge?

Reply:
I've done both of these things, as well as:

- use a new variadic method that looks nicer
- refactor to move the static merge function from `Structure` to `EvaluationContext`

see: https://github.com/open-feature/java-sdk/pull/1178/commits/f3e06863c60d65601ac41a6421b1c0797bde6783